### PR TITLE
Fix/adf 485/criteria dropdown slowness fix 2

### PR DESCRIPTION
**Pull request by gabrielfs7**
Optmize indexation and tree build for metadata: 

https://oat-sa.atlassian.net/browse/ADF-225

### CODE REVIEW

File: model/Metadata/Service/ClassMetadataSearcher.php [hunk 1]
```diff
@@ -52,59 +49,102 @@ public function findAll(ClassMetadataSearchInput $input): ClassCollection
...
 
             return new ClassCollection(...array_values($this->processedClasses));
         }
 
         return $this->getClassMetadataSearcher()->findAll($input);
     }
 
-    private function getSubProperties(string $classUri, array $properties = []): void
+    /**
+     * Given the following class tree relation:
```
Comment (gabrielfs7) [inline model/Metadata/Service/ClassMetadataSearcher.php:61]: I am against comment unless they are really necessary. This was a case IMO :)

**Review by boajer — APPROVED** [review]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful

**Review by chinnu-focaloid — APPROVED** [review]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [X] New code is respecting code style rules
- [X] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [] Feature is working correctly on my local machine (if applicable)
- [X] Acceptance criteria are respected
- [X] Pull request title and description are meaningful

File: model/Metadata/Factory/ClassPathFactory.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            if ($this->isRootClass($parentClass)) {
+                $path[] = $parentClass->getUri();
+
+                break;
+            }
+
+            $path[] = $parentClass->getUri();
+        }
+
+        return $path;
```
Comment (shpran) [inline model/Metadata/Factory/ClassPathFactory.php:52]: Do we need to check root class on every iteration?

```suggestion
        $path = [$class->getUri()];
        
        if ($this->isRootClass($class)) {
            return $path;
        }

        foreach ($class->getParentClasses(true) as $parentClass) {
            if ($this->isRootClass($parentClass)) {
                $path[] = $parentClass->getUri();

                break;
            }

            $path[] = $parentClass->getUri();
        }

        return $path;
```

File: model/Metadata/Listener/MetadataInheritanceListener.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+use oat\taoAdvancedSearch\model\Index\Service\ResultIndexer;
+use oat\taoAdvancedSearch\model\Metadata\Normalizer\MetadataNormalizer;
+
+class MetadataInheritanceListener extends ConfigurableService implements ListenerInterface
+{
+    public const SERVICE_ID = 'taoAdvancedSearch/MetadataInheritanceListener';
+
+    /**
+     * @throws UnsupportedEventException
+     */
```
Comment (shpran) [inline model/Metadata/Listener/MetadataInheritanceListener.php:39]:
```suggestion
```

**Review by shpran — APPROVED** [review]
- [x] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful

File: model/Metadata/Factory/ClassPathFactory.php (hunk 2 above)
Comment (gabrielfs7) [reply model/Metadata/Factory/ClassPathFactory.php:52]: Good point! Will improve, thanks!

Comment (gabrielfs7) [reply model/Metadata/Factory/ClassPathFactory.php]: Improved 👍

**Comment by AntonNeviarouski** [comment]
QA verified on http://test-advanced-criteria.playground.kitchen.it.taocloud.org:42161/

**Pull request merged**